controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java

```diff
@@ -512,12 +512,7 @@ public void listScopes(final SecurityContext securityContext, final AsyncRespons
                          .thenApply(scopesList -> {
                              ScopesList scopes = new ScopesList();
                              scopesList.forEach(scope -> {
-                                 try {
-                                     authorize(scope, principal, READ);
```

Comment:
I understand that removing this fixes the issue where someone has READ on / (and therefore shouldn't be required to have READ on individual scopes underneath).

But what if there are 10 scopes, and I have read on scope number 3.   When I do list on /, I should get a list back with just scope number 3 in it, I believe?

With the current code:   I would get access denied because I don't have READ on / 
With the proposes changes:   Same.

(I guess this PR doesn't make things worse and at least fixes the issue for the use case when I have READ on /.   Just pointing out if there's another problem along side of it?)

Reply:
So maybe the ultimate fix is:  look at "/" first,  and if not authorized there, don't bail, look at individual scope permissions?

Reply:
This is the current implementation. Please see this: https://github.com/pravega/pravega/pull/3110/files/3f482a46aaa92dcf2c1330d82312ebb6049b2044#diff-7d4abaf3b0a8ec424c0bec59a93cf94dL503. ListScopes will list all the scopes in the scenario that you mentioned.

Reply:
But if I don't have READ on /.  Just on scope 3, what happens?

Reply:
My feeling is that `ListScopes` should list all the scopes that one has access to, or an empty list if one has access to no scopes.   Meanwhile, it makes sense that one may be granted access to all scopes, or granted access only to specific scopes.    

@arvindkandhare don't you agree that ListScopes should filter its results?

Reply:
The analogy with the filesystem does make sense.     But as @fpj explains, it's also true, we'll need to be able to get a filtered list of "only what you can see".   So a UI application wouldn't present scopes to a user which would only cause problems later on because the user doesn't have permissions on it for example.

The PR as it stands, seems to solve the issue for the user that has, exclusively, READ on /.   (a use case that's currently broken).   So it does solve that and has value on its own.

It would be reasonable IMO to tackle the filtering in a separate PR, although, the reason why I initially brought it up is because it looked like all the pieces to do both were right there.  The only thing needed was to not fail if the top level "/" check fails authorization, but give a chance to the subsequent logic below to also look at individual permissions to add in "allowed" scopes.

Basically "/" being a wild card that trumps the rest.

Reply:
Ok, let's try to agree on the approach to follow here. The options I see are:

1- Stick to the file system model and keep the change as is. I'd still like to add a test case. but otherwise keep the change as is.
2- Filter according to authorization. It requires a change, but it is a trivial one. I'm on the fence about exposing scope names to all users. It doesn't seem to be a big deal as an unauthorized user can't see the streams in the scope or access stream data.
3- Create an additional call to return only authorized scopes to make it simpler for presentation purposes (e.g., in a UI).

Reply:
I think that leaving "as is" (option 1) is not bad, but is incomplete as it doesn't allow users that only have access to a few scopes, to "list [their] scopes".

Regarding (2), I agree that from a security perspective, it's not a big deal to show scopes they don't have access to because anyway they can't even do a get on these individual scopes or access data.  But, it misguides UIs etc., in allowing users to "select" some scope they may have gotten from the list, and later on have authz issues due to the fact they don't actually have access.

For (3), I feel it would just complicate things to have 2 different calls.  One might say: if I have access on "/" and I use this new filtered scopes list call, what would it show?    Having a list-all and list-just-what-i-can-see feels like it could confuse people (maybe).

To have it all, it seems like option 2, with the filtering, is the way to go.   If a user has READ on /, we return all, if not, we get down to checking individual READ on each scope to produce the filtered list.  (that would be my vote).    Though I do see the appeal of the filesystem analogy, my humble opinion is that the one call that does everything we need trumps it.

Reply:
Say that for administration purposes, you do want to know all the scopes, would we simply give at least read access to all scopes to such an admin user?

Reply:
I think an admin would be given READ on "/".

Reply:
TLDR; trim the results of ListScopes.

My full opinion:
1. any resemblance to a file system is superficial; focus on the logical security model that you want.
2. scopes are the foundation for isolation, are likely to have quota or similar policy in the future, and so we should expect the typical user to have access to few scopes.
3. having read access to the admin scopes should be considered a very high level of access
4. it is typical in multi-tenant cloud services that each tenant is 'a world unto themselves'; little-to-no visibility of other tenants.   Enumerating all scopes seems at odds with cloud service principles.